src/main/java/com/spartronics4915/frc2019/subsystems/CargoChute.java

```diff
@@ -95,52 +111,50 @@ public void onLoop(double timestamp)
                 switch (mSystemState)
                 {
-                    case RAMPING:
+                    case INTAKING:
                         if (mStateChanged)
                             mRampMotor.set(ControlMode.PercentOutput, Constants.kRampSpeed);
-                        if (ballInPosition() && !isInManual() && newState == mSystemState)
-                            newState = SystemState.HOLDING;
                         break;
                     case HOLDING:
                         if (mStateChanged)
                             mRampMotor.set(ControlMode.PercentOutput, 0.0);
-                        if (!ballInPosition() && !isInManual() && newState == mSystemState)
-                            newState = SystemState.RAMPING;
                         break;
                     case EJECTING:
                         if (mStateChanged)
                             mRampMotor.set(ControlMode.PercentOutput, -Constants.kRampSpeed);
                         break;
-                    case LOWERING:
-                        if (mStateChanged)
-                            mRampSolenoid.set(Constants.kRampSolenoidRetract);
-                        break;
-                    case RAISING:
-                        if (mStateChanged)
-                            mRampSolenoid.set(Constants.kRampSolenoidExtend);
-                        break;
-                    case SHOOTING_ROCKET:
+                    case SHOOTING:
                         if (mStateChanged)
                         {
-                            mRampSolenoid.set(Constants.kRampSolenoidRetract);
                             mCargoTimer.start();
-                            mRampMotor.set(ControlMode.PercentOutput, Constants.kRampSpeed);
+                            if (mWantedState == WantedState.SHOOT_BAY)
+                                mRampSolenoid.set(Constants.kRampSolenoidExtend);
+                            else
+                                mRampMotor.set(ControlMode.PercentOutput, Constants.kShootSpeed);
                         }
-                        if (mCargoTimer.hasPeriodPassed(Constants.kShootTime) && newState == mSystemState)
-                            newState = SystemState.HOLDING;
+                        if (mCargoTimer.hasPeriodPassed(Constants.kShootTime))
+                        {
+                            mRampSolenoid.set(Constants.kCargoIntakeSolenoidRetract);
+                            mWantedState = WantedState.HOLD;
+                            break; // XXX: is this necessary
+                        }
+                         /* FIXME: problem: below code will run more than once. mStateChanged usually prevents this, but...
+                          * solution: check if the motor is running and replace the mWantedState check with that.
+                          * but .get() doesn't work on mRampMotor, a TalonSRX? i'm not sure why it isn't working */
```

Comment:
`TalonSRX::get` doesn't work because you're just getting the last setpoint. _If_ the Talon had an encoder attached you would be able to use `TalonSRX::getSelectedSensorPosition`. You would _not_ want to just look at velocity, but instead do proactive current limiting [(you use some prior knowledge about the motor and its output shaft velocity)](https://spartronics.slack.com/archives/G2YP9RD7D/p1550705056023500?thread_ts=1550692335.009800&cid=G2YP9RD7D).

We do not, however, have an encoder attached, so you should use the TalonSRX current limiting mode. @binnur [linked to on Slack](https://spartronics.slack.com/archives/G2YP9RD7D/p1550707123025500?thread_ts=1550692335.009800&cid=G2YP9RD7D).

Reply:
Alright. For now, I'm using a boolean to make sure it only runs once.